omp/btx_ompinterval_callbacks.cpp

```diff
@@ -0,0 +1,139 @@
+#include "magic_enum.hpp"
+#include "ompt.h.include"
+#include "xprof_utils.hpp"
+#include <iostream>
+#include <metababel/metababel.h>
+#include <string>
+#include <unordered_map>
+
+using hpt_function_name_omp_t =
+    std::tuple<hostname_t, process_id_t, thread_id_t, std::string>;
+
+struct data_s {
+  std::unordered_map<hpt_function_name_omp_t, uint64_t> host_start;
+};
+typedef struct data_s data_t;
+
+static void btx_initialize_component(void **usr_data) {
+  *usr_data = new data_t;
+}
+
+static void btx_finalize_component(void *usr_data) {
+  delete static_cast<data_t *>(usr_data);
+}
+
+// returns start_ts, for use in traffic or other messages on scope end
+static int64_t _host_op_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                 const char *hostname, int64_t vpid,
+                                 uint64_t vtid, ompt_scope_endpoint_t endpoint,
+                                 std::string const &op_name) {
+  auto state = static_cast<data_t *>(usr_data);
+  if (endpoint == ompt_scope_begin) {
+    state->host_start[{hostname, vpid, vtid, op_name}] = ts;
+  } else {
+    // TODO: detect error?
+    const bool err = false;
+    auto start_ts = state->host_start[{hostname, vpid, vtid, op_name}];
+    if (start_ts == 0) {
+      start_ts = ts;
+    }
```

Comment:
Do you know why  we have this check?

Reply:
This is for the cope = beginend case, I will add a comment